nemo/collections/nlp/models/language_modeling/megatron_base_model.py

```diff
@@ -805,19 +816,18 @@ def build_model_parallel_config(self) -> ModelParallelConfig:
 
         return model_parallel_config
 
-    def _prefetch(self, iterator):
-        """Checks if the iterator still has elements to return.
-        Used in models using dataloader_iter to prefetch the next batch before fwd_bwd func
-        is called to avoid PP rank 2 from wait indefinitely to get outpits from PP 1
+    def _val_iterator_done(self, iterator):
+        """
+        Check if we reached trainer.limit_val_batches, if so exhaust the iterator to raise a StopIteration and exit validation_step
         """
-        elements = []
-        num_microbatches = get_num_microbatches()
-        for _ in range(num_microbatches):
+        if (
+            self._val_micro_batches_consumed == self.trainer.num_sanity_val_steps
+            or self._val_micro_batches_consumed == self.trainer.limit_val_batches
+        ):
+            self._val_micro_batches_consumed = 0
             try:
-                element = next(iterator)
-                elements.append(element)
+                _ = next(iterator)  # exhausting the iterator so that PTL knows to go to validation_epoch_end
```

Comment:
Lightning requires to hit a `StopIteration` to break out of the evaluation loop, otherwise it keeps running indefinitely. It hits a StopIteration at the (n+1)th step so we check for `StopIteration` after `limit_val_batches` is exhausted.